tractor/_broadcast.py

```diff
@@ -0,0 +1,308 @@
+    with ``.subscribe()`` and receiving from the new instance it delivers.
+
+    '''
+    def __init__(
+        self,
+
+        rx_chan: AsyncReceiver,
+        state: BroadcastState,
+        receive_afunc: Optional[Callable[[], Awaitable[Any]]] = None,
+
+    ) -> None:
+
+        # register the original underlying (clone)
+        self.key = id(self)
+        self._state = state
+        state.subs[self.key] = -1
+
+        # underlying for this receiver
+        self._rx = rx_chan
+        self._recv = receive_afunc or rx_chan.receive
+        self._closed: bool = False
+
+    async def receive(self) -> ReceiveType:
+
+        key = self.key
+        state = self._state
+
+        # TODO: ideally we can make some way to "lock out" the
+        # underlying receive channel in some way such that if some task
+        # tries to pull from it directly (i.e. one we're unaware of)
+        # then it errors out.
+
+        # only tasks which have entered ``.subscribe()`` can
+        # receive on this broadcaster.
+        try:
+            seq = state.subs[key]
+        except KeyError:
+            if self._closed:
+                raise trio.ClosedResourceError
+
+            raise RuntimeError(
+                f'{self} is not registerd as subscriber')
+
+        # check that task does not already have a value it can receive
+        # immediately and/or that it has lagged.
+        if seq > -1:
+            # get the oldest value we haven't received immediately
+            try:
+                value = state.queue[seq]
+            except IndexError:
+
+                # adhere to ``tokio`` style "lagging":
+                # "Once RecvError::Lagged is returned, the lagging
+                # receiver's position is updated to the oldest value
+                # contained by the channel. The next call to recv will
+                # return this value."
+                # https://tokio-rs.github.io/tokio/doc/tokio/sync/broadcast/index.html#lagging
+
+                # decrement to the last value and expect
+                # consumer to either handle the ``Lagged`` and come back
+                # or bail out on its own (thus un-subscribing)
+                state.subs[key] = state.maxlen - 1
+
+                # this task was overrun by the producer side
+                task: Task = current_task()
+                raise Lagged(f'Task {task.name} was overrun')
+
+            state.subs[key] -= 1
+            return value
+
+        # current task already has the latest value **and** is the
+        # first task to begin waiting for a new one
+        if state.recv_ready is None:
+
+            if self._closed:
+                raise trio.ClosedResourceError
+
+            event = trio.Event()
+            state.recv_ready = key, event
+
+            # if we're cancelled here it should be
+            # fine to bail without affecting any other consumers
+            # right?
+            try:
+                value = await self._recv()
```

Comment:
My last edge case concern in the one where this call gets cancelled and we leave the `state.recv_ready` event consumers potentially stuck. What should happen then?

If we wake them up then we need one of them to re-enter and make this call again.
What happens to the others?
They have to be placed back in the event `.wait()`

In theory the below commented recursive call would solve this right?
Is it necessary tho?